File: src/commonMain/kotlin/org/agraman4u/appconfig/AppConfig.kt
package org.agraman4u.appconfig

import org.agraman4u.appconfig.parser.AppConfigParser
import org.agraman4u.appconfig.parser.Parser
import org.agraman4u.appconfig.utils.AppConfigurationArgs
import org.agraman4u.appconfig.utils.getRegion
import org.agraman4u.appconfig.utils.getStage

class AppConfig private constructor(args: AppConfigurationArgs) {
    val parser: Parser = AppConfigParser(args)

    companion object {
        lateinit var appConfig: AppConfig

        fun initAppConfig(
            serviceName: String,
            stage: String = getStage(),
            region: String = getRegion(),
            configDir: String? = null
        ) {
            appConfig = AppConfig(AppConfigurationArgs(stage, region, serviceName, configDir))
        }

        inline fun <reified T> get(identifier: String, key: String): T {
            val value = appConfig.parser.getValue(identifier, key) as String
            return when (T::class) {
                Int::class -> value.toInt() as T
                Long::class -> value.toLong() as T
                Float::class -> value.toFloat() as T
                Double::class -> value.toDouble() as T
                String::class -> value as T
                else -> throw IllegalStateException("Type ${T::class.simpleName} not supported to be parse")
            }
        }

        fun getAllEntries(identifier: String): Map<String, String> {
            return appConfig.parser.getAllEntries(identifier) ?: emptyMap()
        }
    }
}

File: src/app_config_parser/app_config.py
from typing import TypeVar, Dict, Optional, Type
from app_config_parser.parser.app_config_parser import AppConfigParser
from app_config_parser.parser.parser import Parser
from app_config_parser.utils.types import AppConfigurationArgs
from app_config_parser.utils.properties import get_stage, get_region

T = TypeVar('T')

class AppConfig:
    _app_config: Optional['AppConfig'] = None
    
    def __init__(self, args: AppConfigurationArgs):
        self.parser: Parser = AppConfigParser(args)
    
    @classmethod
    def init_app_config(cls, service_name: str, stage: str = None, region: str = None):
        if stage is None:
            stage = get_stage()
        if region is None:
            region = get_region()
        
        cls._app_config = cls(AppConfigurationArgs(stage, region, service_name))
    
    @classmethod
    def get(cls, identifier: str, key: str, return_type: Type[T] = str) -> T:
        if cls._app_config is None:
            raise RuntimeError("AppConfig not initialized. Call init_app_config() first.")
        
        value = cls._app_config.parser.get_value(identifier, key)
        if value is None:
            raise KeyError(f"Key '{key}' not found in identifier '{identifier}'")
        
        if return_type == str:
            return value
        elif return_type == int:
            return int(value)
        elif return_type == float:
            return float(value)
        elif return_type == bool:
            return value.lower() in ('true', '1', 'yes', 'on')
        else:
            raise ValueError(f"Unsupported type: {return_type}")
    
    @classmethod
    def get_all_entries(cls, identifier: str) -> Dict[str, str]:
        if cls._app_config is None:
            raise RuntimeError("AppConfig not initialized. Call init_app_config() first.")
        
        return cls._app_config.parser.get_all_entries(identifier) or {}